## Table of contents
## 1 Importing libraries and data files
## 2 If-Statement with User-Defined Functions
## 3 Creating the variable "price label"
## 4 Creating the variable "busiest day"
## 5 Task
### Question 2: Creating the variable "busiest days"
### Question 3: Accuracy checks
### Question 4: Create a new column called “busiest_period_of_day” containing labels “Most orders,” “Average orders,” and “Fewest orders”
### Question 5: Frequencies of the new column/variable
### Question 6: Exporting files

## 1 Importing libraries and data files

In [1]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Importing data files

path = r'C:\Users\Eva\Documents\Instacart Basket Analysis'
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merged.pkl'))

## 2 If-Statement with User-Defined Functions

In [9]:
# Creating a subset

df_subset = df_ords_prods_merged[:1000000]

In [10]:
df_subset.shape

(1000000, 14)

In [11]:
# Defining the price label function

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [13]:
# Applying the price label function

df_subset['price_range'] = df.apply(price_label, axis=1)

MemoryError: 

## 3 Creating the variable "price label"

In [5]:
#If-statements with the loc-function

df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_label'] = 'High-range product'

In [6]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_label'] = 'Mid-range product' 

In [7]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_label'] = 'Low-range product'

In [8]:
# Checking the frequencies

df_ords_prods_merged['price_label'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_label, dtype: int64

## 4 Creating the variable "busiest day"

In [9]:
# Checking the frequencies

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)


0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [10]:
# Creating an empty list

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [11]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [12]:
df_ords_prods_merged['busiest_day'] = result

In [13]:
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

## 5 Task 

### Question 2: Creating the variable "busiest days"

In [14]:
# Creating an empty list refering to the frequencies of "orders_day_of_week" from above: 
# Saturday (=0) and Sunday (=1) are the busiest days, Wednesday (=4) and Tuesday (=3) are the slowest days. 
# Every other day is regularly busy.

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest days")
  elif value == 1:
    result.append("Busiest days")
  elif value == 4:
    result.append("Slowest days")
  elif value == 3:
    result.append("Slowest days")
  else:
    result.append("Regularly busy")

In [15]:
result

['Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Slowest days',
 'Regularly busy',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Slowest days',
 'Slowest days',
 'Regularly busy',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly busy',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

In [16]:
# Creating the new variable

df_ords_prods_merged['busiest_day_2'] = result

### Question 3: Accuracy checks

In [17]:
# Accuracy/frequency checks

df_ords_prods_merged['busiest_day_2'].value_counts(dropna = False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_day_2, dtype: int64

In [18]:
 df_ords_prods_merged['busiest_day_2'].describe()

count           32404859
unique                 3
top       Regularly busy
freq            12916111
Name: busiest_day_2, dtype: object

Nothing unusual spotted, everything seems to be as expected. "Regulaly busy" is the most frequent label (followed by "Busiest days" and "Slowest days"). Also, the values add up correctly to the overall count of 32404859 rows. 

### Question 4: Create a new column called “busiest_period_of_day” containing labels “Most orders,” “Average orders,” and “Fewest orders”

In [19]:
# Checking the frequencies

df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [20]:
# Creating an empty list refering to the frequencies of "order_hour_of_day" from above:
# The top 8 hours of the day (in terms of highest frequencies of orders) get the label "Most orders", 
# the bottom 8 hours of the day get the label "Fewest orders" 
# and the 8 hours of the day in between the two get the label "Average orders".

result = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value == 3:
    result.append("Fewest orders")
  elif value == 4:
    result.append("Fewest orders")  
  elif value == 2:
    result.append("Fewest orders")  
  elif value == 5:
    result.append("Fewest orders")  
  elif value == 1:
    result.append("Fewest orders")
  elif value == 0:
    result.append("Fewest orders")
  elif value == 6:
    result.append("Fewest orders")
  elif value == 23:
    result.append("Fewest orders")
  elif value == 9:
    result.append("Most orders")
  elif value == 16:
    result.append("Most orders")
  elif value == 12:
    result.append("Most orders")
  elif value == 13:
    result.append("Most orders")
  elif value == 15:
    result.append("Most orders")
  elif value == 14:
    result.append("Most orders")
  elif value == 11:
    result.append("Most orders")
  elif value == 10:
    result.append("Most orders")
  else:
    result.append("Average orders")

In [21]:
result

['Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most ord

In [22]:
# Creating the new variable

df_ords_prods_merged['busiest_hour_of_day'] = result

### Question 5: Frequencies of the new column/variable

In [23]:
# Frequency checks

df_ords_prods_merged['busiest_hour_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: busiest_hour_of_day, dtype: int64

In [24]:
df_ords_prods_merged['busiest_hour_of_day'].describe()

count        32404859
unique              3
top       Most orders
freq         21118071
Name: busiest_hour_of_day, dtype: object

### Question 6: Exporting files

In [25]:
df_ords_prods_merged.to_csv(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merged_2.csv'))